File: MathCore.DSP/Filters/EllipticBandStop.cs
using System.Data;
// ReSharper disable InconsistentNaming

namespace MathCore.DSP.Filters;

/// <summary>Эллиптический полосо-заграждающий фильтр</summary>
public class EllipticBandStop : EllipticFilter
{
    /// <summary>Формирование спецификации фильтра</summary>
    /// <param name="dt">Период дискретизации сигнала</param>
    /// <param name="fpl">Частота нижней границы полосы пропускания</param>
    /// <param name="fsl">Частота нижней границы полосы заграждения</param>
    /// <param name="fsh">Частота верхней границы полосы заграждения</param>
    /// <param name="fph">Частота верхней границы полосы пропускания</param>
    /// <param name="Gp">Уровень АЧХ в полосе пропускания (в единицах 0..1)</param>
    /// <param name="Gs">Неравномерность АЧХ в полосе заграждения (в единицах 0..1)</param>
    /// <returns>Спецификация фильтра</returns>
    private static Specification GetSpecification(
        double dt,
        double fpl,
        double fsl,
        double fsh,
        double fph,
        double Gp,
        double Gs)
    {
        if (Gp <= Gs)
            throw new ArgumentOutOfRangeException(
                nameof(Gp), Gp, $"Уровень АЧХ в полосе пропускания Gp={Gp} был меньше, либо равен уровню АЧХ в полосе заграждения Gs={Gs}");

        var Fpl = ToDigitalFrequency(fpl, dt);
        var Fsl = ToDigitalFrequency(fsl, dt);
        var Fsh = ToDigitalFrequency(fsh, dt);
        var Fph = ToDigitalFrequency(fph, dt);

        var Wpl = Consts.pi2 * Fpl;
        var Wsl = Consts.pi2 * Fsl;
        var Wsh = Consts.pi2 * Fsh;
        var Wph = Consts.pi2 * Fph;

        var Wc = Wsl * Wsh;
        var dW = Wsh - Wsl;

        // Выбор опорной частоты
        // Если   Wc / Wph > Wpl
        // то есть      Wc > Wpl*Wph
        // то есть Wsl*Wsh > Wpl*Wph
        // то есть центральная частота по границам подавления > центральной частоты по границам пропускания
        // то выбираем в качестве опорной частоты выбираем верхнюю границу пропускания
        // иначе, выбираем нижнюю границу пропускания
        var Wp = Wc / Wph > Wpl
            ? Wph
            : Wpl;
        var Fp = Math.Abs(dW * Wp / (Wc - Wp.Pow2())) / Consts.pi2;
        const double Fs = 1 / Consts.pi2;

        // Для передачи информации о граничных частотах в спецификацию аналогвого прототипа перечситываем частоты цифрового фильтра обратно
        var fp = ToAnalogFrequency(Fp, dt);
        var fs = ToAnalogFrequency(Fs, dt);

        return new Specification(dt, fp, fs, Gp, Gs);
    }

    /// <summary>Расчёт коэффициентов полиномов числителя из знаменателя передаточной функции фильтра</summary>
    /// <param name="fsl">Нижняя частота полосы подавления</param>
    /// <param name="fsh">Верхняя частота полосы подавления</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <returns>Кортеж, содержащий массивы A - коэффициенты полинома знаменателя и B - коэффициенты полинома числителя</returns>
    private static (double[] A, double[] B) Initialize(double fsl, double fsh, Specification Spec)
    {
        var kW = 1 / Spec.kW;
        var kEps = 1 / Spec.kEps;

        var N = (int)Math.Ceiling(T(kEps) * K(kW) / (K(kEps) * T(kW))); // Порядок фильтра

        // Нули и полюса аналогового прототипа - эллиптического ФНЧ
        var (zeros, poles) = GetNormedZerosPoles(N, Spec.EpsP, Spec.EpsS, Spec.Wp);

        // Пересчитываем аналоговые частоты полосы заграждения в цифровые
        var dt = Spec.dt;
        var Fsl = ToDigitalFrequency(fsl, dt);
        var Fsh = ToDigitalFrequency(fsh, dt);

        // Переносим нули и полюса аналогового нормированного ФНЧ в полосы ПЗФ
        var pzf_zeros = TransformToBandStop(zeros, Fsl, Fsh);
        var pzf_poles = TransformToBandStop(poles, Fsl, Fsh);

        // Если фильтр нечётный, то надо добавить ещё одну пару нулей на центральной частоте полосы заграждения
        if (N.IsOdd())
        {
            var wc_sqrt = Consts.pi2 * (fsl * fsh).Sqrt(); // странно что не Fsl и Fsh
            pzf_zeros = pzf_zeros.AppendLast(
                (0, +wc_sqrt),
                (0, -wc_sqrt));
        }

        // Преобразуем аналоговые нули и полюса в нули и полюса цифрового фильтра с помощью Z-преобразования
        var z_zeros = ToZArray(pzf_zeros, dt);
        var z_poles = ToZArray(pzf_poles, dt);

        // Вычисляем коэффициент нормировки фильтра на нулевой частоте 
        var g_norm = (N.IsOdd() ? 1 : Spec.Gp)
            / (z_zeros.Multiply(z => 1 - z) / z_poles.Multiply(z => 1 - z)).Abs;

        // Определяем массивы нулей коэффициентов полиномов знаменателя и числителя
        var B = Polynom.Array.GetCoefficientsInverted(z_zeros).ToArray(b => b.Re * g_norm);
        var A = Polynom.Array.GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="dt">Период дискретизации цифрового сигнала</param>
    /// <param name="fpl">Нижняя граница полосы пропускания</param>
    /// <param name="fsl">Нижняя граница полосы подавления</param>
    /// <param name="fsh">Верхняя граница полосы подавления</param>
    /// <param name="fph">Верхняя граница полосы пропускания</param>
    /// <param name="Gp">Уровень АЧХ в полосе пропускания (по умолчанию -1дБ)</param>
    /// <param name="Gs">Уровень АЧХ в полосе подавления (по умолчанию -30дБ)</param>
    public EllipticBandStop(
        double dt,
        double fpl,
        double fsl,
        double fsh,
        double fph,
        double Gp = 0.89125093813374556,
        double Gs = 0.01)
        : this(fsl, fsh, GetSpecification(dt, fpl, fsl, fsh, fph, Gp, Gs)) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="fsl">Нижняя граница полосы подавления</param>
    /// <param name="fsh">Верхняя граница полосы подавления</param>
    /// <param name="Spec">Спецификация фильтра</param>
    private EllipticBandStop(double fsl, double fsh, Specification Spec) : this(Initialize(fsl, fsh, Spec), Spec) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="Polynoms">Кортеж с коэффициентами полиномов знаменателя и числителя функции фильтра</param>
    /// <param name="Spec">Спецификация фильтра</param>
    private EllipticBandStop((double[] A, double[] B) Polynoms, Specification Spec) : this(Polynoms.B, Polynoms.A, Spec) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="B">Коэффициенты полинома числителя</param>
    /// <param name="A">Коэффициенты полинома знаменателя</param>
    /// <param name="Spec">Спецификация фильтра</param>
    private EllipticBandStop(double[] B, double[] A, Specification Spec) : base(B, A, Spec) { }
}

File: MathCore.DSP/Filters/Python/filters/elliptic_bandstop.py
# Полосозаграждающий эллиптический фильтр (порт C# EllipticBandStop)
import numpy as np
from math import ceil, pi, fabs, sqrt

from .elliptic_filter import EllipticFilter
from .analog_based_filter import AnalogBasedFilter, Specification
from .digital_filter import DigitalFilter

pi2 = 2.0 * pi


class EllipticBandStop(EllipticFilter):
    """
    Цифровой полосозаграждающий эллиптический фильтр (ПЗФ).

    Нули в аналоговой области расположены на мнимой оси вблизи центра полосы заграждения.
    Нечётный порядок добавляет дополнительную пару нулей на центральной частоте.
    """

    @staticmethod
    def _get_specification(dt: float, fpl: float, fsl: float, fsh: float, fph: float,
                           Gp: float, Gs: float) -> Specification:
        """Спецификация ПЗФ через эквивалентный ФНЧ-прототип."""
        Fpl = DigitalFilter.to_digital_frequency(fpl, dt)
        Fsl = DigitalFilter.to_digital_frequency(fsl, dt)
        Fsh = DigitalFilter.to_digital_frequency(fsh, dt)
        Fph = DigitalFilter.to_digital_frequency(fph, dt)

        Wpl, Wsl = pi2 * Fpl, pi2 * Fsl
        Wsh, Wph = pi2 * Fsh, pi2 * Fph

        Wc = Wsl * Wsh
        dW = Wsh - Wsl

        Wp = Wph if (Wc / Wph > Wpl) else Wpl
        Fp = fabs(dW * Wp / (Wc - Wp ** 2)) / pi2
        Fs = 1.0 / pi2

        fp = DigitalFilter.to_analog_frequency(Fp, dt)
        fs = DigitalFilter.to_analog_frequency(Fs, dt)
        return Specification(dt, fp, fs, Gp, Gs)

    @staticmethod
    def _initialize(fsl: float, fsh: float, spec: Specification):
        """
        Коэффициенты ПЗФ эллиптического фильтра.

        Алгоритм
        --------
        1. N = ceil(T(1/kEps)·K(1/kW) / (K(1/kEps)·T(1/kW))).
        2. Нули и полюса прototипа масштабируются на W0=Wp.
        3. LP→BS трансформация через цифровые частоты Fsl, Fsh.
        4. Нечётный порядок → дополнительная пара нулей ±j·2π·√(fsl·fsh).
        5. Нормировка по нулевой частоте (DC).

        Примечание: нечётный дополнительный нуль использует исходные fsl, fsh,
        а не предыскажённые Fsl, Fsh — соответствует оригинальному C#-коду.
        """
        K = EllipticFilter.K
        T = EllipticFilter.T

        k_W = 1.0 / spec.kW
        k_eps = 1.0 / spec.kEps

        N = int(ceil(T(k_eps) * K(k_W) / (K(k_eps) * T(k_W))))

        # нули и полюса LP-прototипа масштабированы на Wp
        zeros, poles = EllipticFilter.get_normed_zeros_poles(N, spec.EpsP, spec.EpsS, spec.Wp)

        dt = spec.dt
        Fsl = DigitalFilter.to_digital_frequency(fsl, dt)
        Fsh = DigitalFilter.to_digital_frequency(fsh, dt)

        # LP→BS трансформация через предыскажённые частоты Fsl, Fsh
        pzf_zeros = AnalogBasedFilter.transform_to_band_stop(zeros, Fsl, Fsh)
        pzf_poles = AnalogBasedFilter.transform_to_band_stop(poles, Fsl, Fsh)

        # нечётный порядок: дополнительные нули на центральной частоте полосы заграждения
        # используются исходные fsl, fsh (не Fsl, Fsh) — соответствует C#-коду
        if N % 2 != 0:
            wc_sqrt = pi2 * sqrt(fsl * fsh)
            pzf_zeros = list(pzf_zeros) + [complex(0.0, +wc_sqrt), complex(0.0, -wc_sqrt)]

        z_zeros = DigitalFilter.to_z_array(pzf_zeros, dt)
        z_poles = DigitalFilter.to_z_array(pzf_poles, dt)

        # нормировка по DC (z=1)
        gain_zeros = np.prod([1.0 - z for z in z_zeros])
        gain_poles = np.prod([1.0 - z for z in z_poles])
        k0 = 1.0 if N % 2 != 0 else spec.Gp
        g_norm = k0 / abs(gain_zeros / gain_poles)

        B = np.poly(z_zeros).real * g_norm
        A = np.poly(z_poles).real
        return A, B

    def __init__(self, dt: float, fpl: float, fsl: float, fsh: float, fph: float,
                 Gp: float = 0.891250938, Gs: float = 0.01):
        """
        Создаёт ПЗФ эллиптический.

        Параметры
        ---------
        dt  : период дискретизации (с)
        fpl : нижняя граница полосы пропускания (Гц)
        fsl : нижняя граница полосы заграждения (Гц)
        fsh : верхняя граница полосы заграждения (Гц)
        fph : верхняя граница полосы пропускания (Гц)
        Gp  : коэффициент передачи на границе пропускания
        Gs  : коэффициент передачи на границе заграждения
        """
        spec = EllipticBandStop._get_specification(dt, fpl, fsl, fsh, fph, Gp, Gs)
        A, B = EllipticBandStop._initialize(fsl, fsh, spec)
        super().__init__(B, A, spec)
